SCClassLibrary/Common/GUI/Base/Grid.sc

```diff
@@ -294,18 +294,18 @@ GridLines {
 	var <>spec;
 
 	*new { arg spec;
-		spec = spec.asSpec;
-		spec.warp.switch(
-			LinearWarp(), {
-				^super.newCopyArgs(spec)
-			},
-			ExponentialWarp(), {
-				^ExponentialGridLines(spec)
-			},
-			{
-				^super.newCopyArgs(spec)
-			} // all other Warps
-		)
+		var gridClass;
+		if(spec.isNil) {
+			gridClass = BlankGridLines;
+		} {
+			gridClass = spec.asSpec.warp.gridClass;
+		};
+		if(this.name != \GridLines) {
```

Comment:
Is there a way we could get rid of this test? It looks like something that comes from a lack of clarity in the responsibility. 

So we have one class `GridLines`, which chooses a subclass depending on the argument passed to `*new`. So I suppose the subclasses are not meant to be used in the same way? Then they shouldn't inherit this choosing.

A possible solution would be to use an `AbstractGridLines` class that implements all the common methods and override the `*new` method in `GridLines`.

Sorry for being picky, but let's try :)

Reply:
Okay, I tried :)

Please check the class hierarchy/topology now:
```
GridLines
  AbstractGridLines
    LinearGridLines
    ExponentialGridLines
    BlankGridLines
```
Spec .asWarp calls `LinearGridLines()` or `ExponentialGridLines()` as opposed to GridLines (though maybe that's not needed?)

If this hierarchy is good, I'd also consider moving some implementation from GridLines to AbstractGridLines/

Let me know what you think!

Reply:
I'm also thinking that due to class hierarchy maybe we could use `GridLinesCommon` or `CommonGridLines` etc instead of `AbstractGridLines`?

Reply:
Like @sternsc, I also found it strange that `AbstractGridLines` was not the topmost superclass. I've managed [a solution](https://gist.github.com/mtmccrea/c085627a31c835e5ac1ee80a91287996), I think, so that the taxonomy is:
```
AbstractGridLines
    GridLines
    ExponentialGridLines
    LinearGridLines
        BlankGridLines
```
Carefully note which classes are overriding `*new`, and the change to `AbstractGridLines:-prCheckWarp`.

I tested various specs and grid class combinations with the code below, but it would be good to run it through your tests as well, in particular i haven't thought too deeply about `BlankGridLines`, but seems to work as well.
```
(
c = ControlSpec.new(20,20000, \exp, units: "Hz");
// c = ControlSpec.new(20,2000, \lin, units: "Hz");

// g = GridLines(c);
// g = GridLines(nil);
// g = ExponentialGridLines(c);
g = LinearGridLines(c);

d = DrawGrid.test(g, g, 300@300);
)

d = DrawGrid.test(nil, g, 300@300);
d = DrawGrid.test(g, nil, 300@300);
d = DrawGrid.test(nil, nil, 300@300);
```
EDIT: updated the gist, moving `AbstractGridLines:-getParams` to `LinearGridLines`, making `AbstractGridLines:-getParams` `this.subclassResponsibility`

Reply:
I was thinking about making the hierarchy the way you proposed, but I thought that in such case `Linear/ExponentialGridLines` would not be `.kindOf(GridLines)`, since they are not its subclass. Am I correct to think that? Do you think this is a problem or not?

Reply:
>Linear/ExponentialGridLines would not be .kindOf(GridLines), since they are not its subclass

Conceptually it would make sense if they were subclasses of `GridLines`, but in actual fact they're not, so it probably shouldn't  answer `true`. I did a project-wide search and there are no `.kindOf(GridLines)` tests, so the issue is a hypothetical for existing user code.

Given that `GridLines` returns an instance of a different class, which is atypical, we should specify in the docs that `GridLines` is a convenience class that infers the proper returned type. Hopefully that  would clarify for the user in the (hypothetical) case they ask if the instantiated object is a `GridLines`. 

Maybe other's have thoughts on this?

The main thing I like about the solution i posted is that it's fairly clear that the differences between  the subclasses is largely just the different implementations of `-getParams`, clearly illustrating  how the subclasses differ.

Reply:
I'd usually try to avoid `isKindOf` if possible.

Btw. I am not sure if this is relevant, but `GridLines` need not be a subclass of `AbstractGridLines`, it can just be a factory for `AbstractGridLines` subclasses. So you'd never get an instance of `GridLines` itself.

Or, it might be possible to turn the thing around and use the `Spec` as a factory, in a method `asGridLines`. 

Sorry, I lost the thread a bit, so these comments are just remarks on the side.

Reply:
If you both don't consider the fact that `GridLines` would return a different class (which is not a subclass) as a problematic/breaking API change, then I'd be okay implementing it this way. I'm just wanting to make sure that this change is fine.

Reply:
>Btw. I am not sure if this is relevant, but `GridLines` need not be a subclass of `AbstractGridLines`, it can just be a factory for `AbstractGridLines` subclasses. So you'd never get an instance of `GridLines` itself.

Good point. In fact, at this stage of the redesign that's exactly how `GridLines` behaves, so we should probably remove the `AbstractGridLines` inheritance (updated my [gist](https://gist.github.com/mtmccrea/c085627a31c835e5ac1ee80a91287996) proposal).


>Or, it might be possible to turn the thing around and use the `Spec` as a factory, in a method `asGridLines`.

I'd be in favor of adding this, taking care of its relationship to the `grid` variable. So `ControlSpec`, `Nil`, etc. would have these 3 grid related methods, something like:
(untested)
```supercollider
gridClass { ^this.warp.gridClass }
asGridLines { ^grid = this.warp.gridClass.new(this) }
grid { ^grid ?? { this.asGridLines } }
```
`-gridClass` is then just a convenience, and may not be necessary.

Relatedly, `Nil:-asGrid`, and `(Abstract)GridLines:-asGrid` should probably then be renamed to `asGridLines` (if they existed pre-PR, first added as synonyms then later deprecated?)


>If you both don't consider the fact that `GridLines` would return a different class (which is not a subclass) as a problematic/breaking API change...

I guess I'd be ok with it, provided it's documented, examples updated, etc. 
What do you thinkg @telephon ?

Reply:
I've not followed completely how it will look like, but it seems about right.
I don't think it is a problem to have a factory class return an instance of a different class. But given the `asGridLines` method, we may not even need the factory?

The interface:
```supercollider
gridClass { ^this.warp.gridClass } // yes, this is necessary

// "asSomething" shouldn't change the state of its receiver
asGridLines { ^grid = this.warp.gridClass.new(this) }

// I suppose we don't need this for Nil, 
// and indeed we only need `asGridLines`, right?
grid { ^grid ?? { this.asGridLines } }
```

Reply:
>// "asSomething" shouldn't change the state of its receiver
asGridLines { ^grid = this.warp.gridClass.new(this) }

I agree. I was initially hesitant about that. But considering what the preexisting method does:
```supercollider
grid { ^grid ?? { this.asGridLines } }
```
it's a "getter" which returns the `grid` variable, or creates a `GridLines` _without setting the grid variable_. If you alter the returned object, that change would only be reflected in `ControlSpec`'s "`grid`" in the non-`nil` case, i.e. if it was already set externally (`grid` is never set anywhere in the class). So it's a confused implementation, IMO. 

In any case, I think you're right that it's best to play by the rules and stick with:
`asGridLines { ^this.warp.gridClass.new(this) } // or just this.gridClass.new if we keep the -gridClass method`

>gridClass { ^this.warp.gridClass } // yes, this is necessary

I'm ok with leaving this, but why is this necessary? The grid class corresponding to the spec actually lives in `Warp:-gridClass`.  It was added in this PR in service of creating a `GridLines` from the spec, which would now be done with `asGridLines`. Am I missing another use case?

How comfortable are we with being more aggressive with breaking changes here? I think we're playing around in a fairly lonely corner of the library.
Much of the awkwardness here is from `GridLines` itself, which needs to stay for backwards compatibility, as I understand it. It's essentially been superseded by `AbstractGridLines` and takes on the  new role as a factory. For example:
>// I suppose we don't need this for Nil, 
// and indeed we only need `asGridLines`, right?
grid { ^grid ?? { this.asGridLines } }

yes, this kind of thing could be removed (and ideally similar methods renamed), but it existed pre-PR.

Reply:
I've changed the implementation to use AbstractGridLines as the parent class and GridLines as the factory.

> But given the asGridLines method, we may not even need the factory (GridLines)?

@telephon At the very least we need GridLines for backwards compatibility I think?

> So ControlSpec, Nil, etc. would have these 3 grid related methods, something like:

@mtmccrea for now I kept the implementation that uses `.gridClass`. I guess we could presumably switch to `.asGridLines`, should I go ahead and do that? I don't have a strong preference on this, the "current" implementation (`.gridClass`) looks fine as well.

Reply:
FWIW I [tried implementing](https://github.com/supercollider/supercollider/compare/develop...dyfer:supercollider:GridLines-exp-asGridLines) `.asGridLines`.
The problem is that it didn't allow me to drop the `.gridClass` method. I was able to get the proper GridLines using .asGridLines, but to perform the check that warns the user in case the GridLines they used is not the one returned from the Warp, I'd need to ask the warp for the appropriate `GridLines` _instance_ again, at which point we enter an infinite loop...

Reply:
In light of the above, I see the following possibilities:
- keep the implementation as it is now in [86d0af2](https://github.com/supercollider/supercollider/pull/5161/commits/86d0af24b768b7fb08c8bcf46de0ff7be10c88a6) (we leave .grid/.asGrid as they were, add .gridClass)
- add .asGridLines AND ALSO add .gridClass, deprecate .grid/.asGrid
- add .asGridLines,  deprecate .grid/.asGrid, and don't warn if the explicitly used GridLines is inconsistent with the one that warp would return

Reply:
> @telephon At the very least we need GridLines for backwards compatibility I think?

yes. If it is not too messy, we may keep an interface around. But I would be confident that deprecating the old interface would be ok.

I've looked through the current implementation, this looks very good to me. Are there any open issues with it?


@mtmccrea 
> > // "asSomething" shouldn't change the state of its receiver
> > asGridLines { ^grid = this.warp.gridClass.new(this) }
> 
> I agree. I was initially hesitant about that. But considering what the preexisting method does:
> 
> ```supercollider
> grid { ^grid ?? { this.asGridLines } }
> ```

It is standard that a method like `grid` may change the internal state, especially in cases of lazy initialisation, we have this a lot. Just not a method that starts with `as`. But if we can avoid implicit assignments, it's still better!

Reply:
> I've looked through the current implementation, this looks very good to me. Are there any open issues with it?

The question is whether to replace `.grid` and `.asGrid` with the `.asGridLines` method (see [above](https://github.com/supercollider/supercollider/pull/5161#discussion_r980676035))

Reply:
I suppose `ControlSpec` will still have a settable `grid` instance variable? Then the message `.grid` should be kept. The conversion method `asGrid` seems fine to me. I see no need why it would have to be renamed to `asGridLines` – we simply say that the `GridLines` classes implement grids.

So as long as `asGrid` doesn't modify its receiver, it is fine, I think. `ControlSpec` can implement `asGrid` as well.
Does this make sense to you? I may have overlooked something.

Reply:
Sounds good, I'll keep `.grid` and `.asGrid` as they are now.